### PR TITLE
Add DeepEdit transforms and interaction

### PR DESCRIPTION
**Pull request by diazandr3s**
Signed-off-by: Andres Diaz-Pinto <diazandr3s@gmail.com>

### Description

- Add deepedit transforms currently in MONAI Label repo
- Add Interaction class for multilabel deepedit
- Add TensorBoard Handler to debug multilabel deepedit

These changes help to reduce code in MONAI Label repo and DeepEdit tutorial

### Status
**Ready**

### CODE REVIEW

File: tests/test_deepedit_transforms.py [hunk 1]
```diff
@@ -14,81 +14,303 @@
...
+    SingleLabelSelectiond,
+    SplitPredsLabeld,
+    ToCheckTransformd,
+)
+from monai.utils import min_version, optional_import
 from monai.utils.enums import PostFix
 
-IMAGE = np.array([[[[1, 0, 2, 0, 1], [0, 1, 2, 1, 0], [2, 2, 3, 2, 2], [0, 1, 2, 1, 0], [1, 0, 2, 0, 1]]]])
-LABEL = np.array([[[[0, 0, 0, 0, 0], [0, 1, 0, 1, 0], [0, 0, 1, 0, 0], [0, 1, 0, 1, 0], [0, 0, 0, 0, 0]]]])
+measure, _ = optional_import("skimage.measure", "0.14.2", min_version)
```
Comment (diazandr3s) [inline tests/test_deepedit_transforms.py:34]: @Nic-Ma

Comment (Nic-Ma) [reply tests/test_deepedit_transforms.py:34]: Hi @diazandr3s ,

Please add your test here to skip `min` tests as it has optional import:
https://github.com/Project-MONAI/MONAI/blob/dev/tests/min_tests.py#L168

Thanks.

Comment (diazandr3s) [reply tests/test_deepedit_transforms.py:34]: Many thanks, @Nic-Ma

File: monai/apps/deepedit/transforms.py [hunk 2]
```diff
@@ -11,14 +11,17 @@
 
 import json
 import logging
-from typing import Dict, Hashable, Mapping, Tuple
+import random
+from typing import Dict, Hashable, Mapping, Optional
 
 import numpy as np
+import torch
+from skimage import measure
```
Comment (wyli) [inline monai/apps/deepedit/transforms.py:19]:
```suggestion
measure, _ = optional_import("skimage.measure", "0.14.2", min_version)
```

File: monai/apps/deepedit/transforms.py [hunk 3]
```diff
@@ -56,50 +75,442 @@ def __call__(self, data: Mapping[Hashable, np.ndarray]) -> Dict[Hashable, np.nda
...
+        if sid is not None and dimensions == 3:
+            dims = 2
+            label = label[0][..., sid][np.newaxis]  # Assume channel is first and depth is last CHWD
+
+        # import matplotlib.pyplot as plt
+        # plt.imshow(label[0])
+        # plt.title(f'label as is {key_label}')
+        # plt.show()
+        # plt.close()
+
```
Comment (wyli) [inline monai/apps/deepedit/transforms.py:335]:
```suggestion
```

File: monai/apps/deepedit/transforms.py [hunk 4]
```diff
@@ -56,50 +75,442 @@ def __call__(self, data: Mapping[Hashable, np.ndarray]) -> Dict[Hashable, np.nda
...
+                    pos_guidance.append(self.default_guidance)
+                    continue
+
+            # plt.imshow(label[0])
+            # plt.title(f'Label postprocessed with blob number {key_label}')
+            # plt.show()
+
+            # plt.imshow(distance_transform_cdt(label)[0])
+            # plt.title(f'Transform CDT {key_label}')
+            # plt.show()
```
Comment (wyli) [inline monai/apps/deepedit/transforms.py:363]:
```suggestion
```

File: monai/apps/deepedit/transforms.py [hunk 5]
```diff
@@ -56,50 +75,442 @@ def __call__(self, data: Mapping[Hashable, np.ndarray]) -> Dict[Hashable, np.nda
...
+                        self._apply(tmp_label, self.sid.get(key_label, None), key_label).astype(int).tolist()
+                    )
+                d[self.guidance] = label_guidances
+                return d
+            else:
+                print("This transform only applies to label key")
+        return d
+
+
+class FindDiscrepancyRegionsCustomd(MapTransform):
```
Comment (wyli) [inline monai/apps/deepedit/transforms.py:422]: please rename class

File: monai/apps/deepedit/transforms.py [hunk 6]
```diff
@@ -120,51 +531,401 @@ def find_guidance(self, discrepancy):
...
+        all_guidances = dict()
+        for key_label in self.label_names.keys():
+            clicks = d.get(key_label, [])
+            clicks = list(np.array(clicks).astype(int))
+            all_guidances[key_label] = self._apply(clicks, factor)
+        d[self.guidance] = all_guidances
+        return d
+
+
+class ResizeGuidanceMultipleLabelCustomd(Transform):
```
Comment (wyli) [inline monai/apps/deepedit/transforms.py:687]: please rename

File: monai/apps/deepedit/transforms.py [hunk 7]
```diff
@@ -120,51 +531,401 @@ def find_guidance(self, discrepancy):
...
+                for idx, (key_label, _) in enumerate(d["label_names"].items()):
+                    if key_label != "background":
+                        d[f"pred_{key_label}"] = d[key][idx + 1, ...][None]
+                        d[f"label_{key_label}"] = d["label"][idx + 1, ...][None]
+            elif key != "pred":
+                logger.info("This is only for pred key")
+        return d
+
+
+class ToCheckTransformd(MapTransform):
```
Comment (wyli) [inline monai/apps/deepedit/transforms.py:740]: please consider replacing this one with the existing DataStatsd transform https://docs.monai.io/en/stable/transforms.html#datastats

File: monai/apps/deepedit/transforms.py [hunk 8]
```diff
@@ -120,51 +531,401 @@ def find_guidance(self, discrepancy):
...
+        super().__init__(keys, allow_missing_keys)
+
+    def __call__(self, data: Mapping[Hashable, np.ndarray]) -> Dict[Hashable, np.ndarray]:
+        d: Dict = dict(data)
+        for key in self.key_iterator(d):
+            logger.info(f"Printing key shape in ToCheckTransformd: {d[key].shape}")
+        return d
+
+
+# For missing labels in multilabel task
```
Comment (wyli) [inline monai/apps/deepedit/transforms.py:756]:
```suggestion
```

File: monai/apps/deepedit/transforms.py [hunk 9]
```diff
@@ -120,51 +531,401 @@ def find_guidance(self, discrepancy):
...
+        dims = dimensions
+        if sid is not None and dimensions == 3:
+            dims = 2
+            label = label[0][..., sid][np.newaxis]  # Assume channel is first and depth is last CHWD
+
+        # import matplotlib.pyplot as plt
+        # plt.imshow(label[0])
+        # plt.title(f'label as is {key_label}')
+        # plt.show()
+        # plt.close()
```
Comment (wyli) [inline monai/apps/deepedit/transforms.py:799]:
```suggestion
```

File: monai/apps/deepedit/transforms.py [hunk 10]
```diff
@@ -120,51 +531,401 @@ def find_guidance(self, discrepancy):
...
+
+        label_guidance = []
+        # If there are is presence of that label in this slice
+        if np.max(blobs_labels) <= 0:
+            label_guidance.append(self.default_guidance)
+        else:
+            # plt.imshow(blobs_labels[0])
+            # plt.title(f'Blobs {key_label}')
+            # plt.show()
+            # plt.close()
```
Comment (wyli) [inline monai/apps/deepedit/transforms.py:815]:
```suggestion
```

File: monai/apps/deepedit/transforms.py [hunk 11]
```diff
@@ -120,51 +531,401 @@ def find_guidance(self, discrepancy):
...
+                        continue
+
+                # plt.imshow(label[0])
+                # plt.title(f'Label postprocessed with blob number {key_label}')
+                # plt.show()
+
+                # plt.imshow(distance_transform_cdt(label)[0])
+                # plt.title(f'Transform CDT {key_label}')
+                # plt.show()
+
```
Comment (wyli) [inline monai/apps/deepedit/transforms.py:830]:
```suggestion
```

File: tests/test_deepedit_transforms.py [hunk 12]
```diff
@@ -14,81 +14,303 @@
...
+
+np.random.seed(0)
```
Comment (wyli) [inline tests/test_deepedit_transforms.py:36]: please use the setup/teardown API to set the seeds, see https://github.com/Project-MONAI/MONAI/blob/c51d85e7a878ba0a63305fe89e0165f6463d55a2/tests/test_grid_dataset.py#L29-L33

then the test data would be created within the test cases.

File: monai/apps/deepedit/interaction.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+from monai.data import decollate_batch, list_data_collate
+from monai.engines import SupervisedEvaluator, SupervisedTrainer
+from monai.engines.utils import IterationEvents
+from monai.transforms import Compose
+from monai.utils.enums import CommonKeys
+
+
+class Interaction:
+    """
+    Ignite process_function used to introduce interactions (simulation of clicks) for DeepEdit Training/Evaluation.
```
[the remark this answers is not in the document]
Comment (diazandr3s) [reply monai/apps/deepedit/interaction.py:26]: Agree!

File: monai/apps/deepedit/interaction.py [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+        self,
+        deepgrow_probability: float,
+        transforms: Union[Sequence[Callable], Callable],
+        train: bool,
+        label_names: Dict[str, int],
+        click_probability_key: str = "probability",
+    ) -> None:
+
+        if not isinstance(transforms, Compose):
+            transforms = Compose(transforms)
```
[the remark this answers is not in the document]
Comment (diazandr3s) [reply monai/apps/deepedit/interaction.py:47]: Thanks!

File: monai/apps/deepedit/transforms.py [hunk 15]
```diff
@@ -27,23 +31,38 @@
...
+        allow_missing_keys: bool = False,
+    ):
         """
-        Discard positive and negative points randomly or Add the two channels for inference time
+        Discard positive and negative points according to discard probability
 
-        :param probability: Discard probability; For inference it will be always 1.0
+        :param keys: The ``keys`` parameter will be used to get and set the actual data item to transform
+        :param number_intensity_ch: number of intensity channels
+        :param probability: probability of discarding clicks
```
[the remark this answers is not in the document]
Comment (diazandr3s) [reply monai/apps/deepedit/transforms.py:47]: Thanks!

File: tests/test_deepedit_transforms.py (hunk 12 above)
Comment (diazandr3s) [reply tests/test_deepedit_transforms.py:36]: Please check the updated version

File: monai/apps/deepedit/transforms.py [hunk 16]
```diff
@@ -56,50 +75,442 @@ def __call__(self, data: Mapping[Hashable, np.ndarray]) -> Dict[Hashable, np.nda
...
+                    if key_label != "background":
+                        new_label_names[key_label] = idx
+                        label[d[key] == val_label] = idx
+                    if key_label == "background":
+                        new_label_names["background"] = 0
+
+                d["label_names"] = new_label_names
+                d[key] = label
+            else:
+                print("This transform only applies to the label")
```
[the remark this answers is not in the document]
Comment (diazandr3s) [reply monai/apps/deepedit/transforms.py:110]: Thanks!

File: monai/apps/deepedit/transforms.py [hunk 17]
```diff
@@ -56,50 +75,442 @@ def __call__(self, data: Mapping[Hashable, np.ndarray]) -> Dict[Hashable, np.nda
...
+                # Convert label to index values following label_names argument
+                max_label_val = self.label_names.index(t_label) + 1
+                d[key][d[key] > 0] = max_label_val
+                print(f"Using label {t_label} with number: {d[key].max()}")
+            else:
+                print("This transform only applies to the label")
+        return d
+
+
+class AddGuidanceSignalCustomd(MapTransform):
```
[the remark this answers is not in the document]
Comment (diazandr3s) [reply monai/apps/deepedit/transforms.py:161]: Agree!

File: monai/apps/deepedit/transforms.py (hunk 7 above)
Comment (diazandr3s) [reply monai/apps/deepedit/transforms.py:744]: Agree! Thanks for this
